zcash_transparent/src/zip48.rs

```diff
@@ -0,0 +1,450 @@
+    pub fn key_info_expression<P: consensus::Parameters>(&self, params: &P) -> String {
+        self.key_expression_inner(pub_prefix(params), vec![])
+            .to_string()
+    }
+
+    /// Produces a [`KeyExpression`] from this public key corresponding to a specific
+    /// address.
+    pub fn key_expression_for_address(
+        &self,
+        prefix: Prefix,
+        scope: zip32::Scope,
+        address_index: NonHardenedChildIndex,
+    ) -> KeyExpression {
+        self.key_expression_inner(
+            prefix,
+            vec![
+                TransparentKeyScope::from(scope).into(),
+                address_index.into(),
+            ],
+        )
+    }
+
+    fn key_expression_inner(
+        &self,
+        prefix: Prefix,
+        child: Vec<bip32::ChildNumber>,
+    ) -> KeyExpression {
+        KeyExpression::from_xpub(Some(self.origin.clone()), prefix, self.key.clone(), child)
+            .expect("correct by construction")
+    }
+}
+
+/// A [ZIP 48] P2SH multisig full viewing key.
+///
+/// This provides the necessary derivation capability to view all funds controlled by a
+/// P2SH multisig account.
+///
+/// [ZIP 48]: https://zips.z.cash/zip-0048
+pub struct FullViewingKey {
+    threshold: u8,
+    key_info: Vec<AccountPubKey>,
+}
+
+impl FullViewingKey {
+    /// Constructs a full viewing key for a standard ZIP 48 account.
+    ///
+    /// This uses the standard wallet descriptor template. For example:
+    /// - `threshold == 2`
+    /// - `key_info.len() == 3`
+    /// - Wallet descriptor template: `"sh(sortedmulti(2,@0/**,@1/**,@2/**))"`
+    pub fn standard(
+        threshold: u8,
+        key_info: Vec<AccountPubKey>,
+    ) -> Result<Self, FullViewingKeyError> {
+        if key_info.is_empty() {
+            Err(FullViewingKeyError::NoPubKeys)
+        } else if key_info.len() > 15 {
+            Err(FullViewingKeyError::TooManyPubKeys)
+        } else if usize::from(threshold) > key_info.len() {
+            Err(FullViewingKeyError::InvalidThreshold)
+        } else {
+            // Verify `key_info` in a scope so we can borrow from it.
+            {
+                // To be compatible with ZIP 48, all keys must have the same derivation
+                // information. We check this by collecting into a set, and then checking
+                // it contains one entry. We don't need to verify the derivation's
+                // structure because `AccountPubKey` enforces it by construction.
+                let derivations = key_info
+                    .iter()
+                    .map(|key| key.origin.derivation())
+                    .collect::<BTreeSet<_>>();
+                if derivations.len() != 1 {
+                    return Err(FullViewingKeyError::IncompatiblePubKeys);
+                }
+            }
+
+            // TODO: Decide whether `key_info` should be sorted (or checked to be sorted)
+            // to ensure a canonical multipath descriptor.
```

Comment:
Do we ever expect this type to be constructed via parsing? If so, I think that we'll need to preserve the order of `key_info` so that the parsing will round-trip.

Reply:
Right, that's why I raised this as a question. For now we can avoid it, and kick parsing to the layer above (in my case, `zcash-devtool`).